include/multipass/mount_handler.h

```diff
@@ -18,58 +18,66 @@
 #ifndef MULTIPASS_MOUNT_HANDLER_H
 #define MULTIPASS_MOUNT_HANDLER_H
 
-#include <multipass/disabled_copy_move.h>
-#include <multipass/id_mappings.h>
+#include <multipass/file_ops.h>
 #include <multipass/rpc/multipass.grpc.pb.h>
 #include <multipass/ssh/ssh_key_provider.h>
+#include <multipass/vm_mount.h>
 
 #include <chrono>
-#include <memory>
-#include <string>
 #include <variant>
 
 namespace multipass
 {
 class VirtualMachine;
-struct VMMount;
 
 using ServerVariant = std::variant<grpc::ServerReaderWriterInterface<StartReply, StartRequest>*,
                                    grpc::ServerReaderWriterInterface<LaunchReply, LaunchRequest>*,
                                    grpc::ServerReaderWriterInterface<MountReply, MountRequest>*,
                                    grpc::ServerReaderWriterInterface<RestartReply, RestartRequest>*>;
 
-class MountHandler : private DisabledCopyMove
+struct MountHandler : private DisabledCopyMove
 {
-public:
     using UPtr = std::unique_ptr<MountHandler>;
 
     virtual ~MountHandler() = default;
 
-    // Used to set up anything host side related
-    virtual void init_mount(VirtualMachine* vm, const std::string& target_path, const VMMount& vm_mount) = 0;
-    virtual void start_mount(VirtualMachine* vm, ServerVariant server, const std::string& target_path,
-                             const std::chrono::milliseconds& timeout = std::chrono::minutes(5)) = 0;
-    virtual void stop_mount(const std::string& instance, const std::string& path) = 0;
-    virtual void stop_all_mounts_for_instance(const std::string& instance) = 0;
-    virtual bool has_instance_already_mounted(const std::string& instance, const std::string& path) const = 0;
+    virtual void start(ServerVariant server, std::chrono::milliseconds timeout = std::chrono::minutes(5)) = 0;
+    virtual void stop() = 0;
```

Comment:
The reason behind `start` is mainly for the QEMU mounts. They are initialized in the constructor, i.e. adding the mount arguments to the QEMU command line, when the instance is stopped. And then, only after the instance has started, the `start` method is called, which does the actual `mount` command. The SSHFS and SMB mounts can happen even when the instance is already running, so for them `start` is not really needed and they could do everything in the constructor. I have an idea of how to allow QEMU mounts to happen as the instance is running, which would allow every mount to only need the constructor, but I'm not entirely sure if it's a good idea, so I'll want to discuss it with all of you, and maybe introduce it in a future re-re-refactoring.

As for `stop`, I don't think we can get rid of it. The reason behind it is that unmounting can fail, and when it does, we want to inform the user of the failure and to not remove the mount. That is why `stop` can throw if anything fails. If we tried to handle unmounting only in the destructor, we wouldn't be able to keep the mount in case of unmount failure. So the closest way to get to RAII that I can think of is to make the `stop` method idempotent, and have the unmount happen either when manually calling `stop` and allowing it to throw, or when the object is dropped and any exception is caught and logged, in cases in which we don't care about preserving the mount on failure.

Reply:
OK, I am happy with that, thanks. Very much interested in you idea for live mounting in QEMU too.